### ISSUE TITLE
Add an API endpoint to retrieve the vehicle composition for a journey

### ISSUE THREAD

**Issue by Bertware** [iRail/iRail#383]
The NMBS added vehicle formations to their train map:
https://twitter.com/SIMPYbe/status/1181918580813762561

![Train composition by NMBS](https://user-images.githubusercontent.com/1094793/66641466-1ad5c480-ec1b-11e9-8f1a-ea3237b86108.png)

It would be interesting to add an endpoint to the iRail API, which allows 3rd party apps to retrieve the details for a train, based on the journey. An example could be `/traincomposition?id=IC537&format=json` . 

Source data comes from `https://trainmapjs.azureedge.net/data/composition/<trainNumber>`:

```
[
  {
    "ptCarFrom": {
      "id": 1088,
      "uicCode": "8893401",
      "fromName": "St-Nicolas - Sint-Niklaas"
    },
    "ptCarTo": {
      "id": 319,
      "uicCode": "8893401",
      "fromName": "Termonde - Dendermonde"
    },
    "confirmedBy": "Atlas",
    "materialUnits": [
      {
        "hasReadingLights": true,
        "hasFoldingSeats": true,
        "hasAssistanceButtons": true,
        "hasToilets": true,
        "hasTables": true,
        "hasSecondClassOutlets": true,
        "hasFirstClassOutlets": true,
        "hasHeating": true,
        "hasAirco": true,
        "typeDescriptionNl": "HV M6 BUH",
        "tractionType": "HV",
        "pullPushType": "R",
        "diagramServiceNumber": 57,
        "materialNumber": 62263,
        "materialDiagram": "KR",
        "uicCode": "508826922637",
        "isLocoInUse": false,
        "inDirection": true,
        "tractionPosition": 2,
        "positionInComposition": 1,
        "materialSubTypeName": "M6BUH",
        "typeDescriptionFr": "HV M6 BUH",
        "hasPriorityPlaces": true,
        "brutoWeight": 64,
        "lengthInMeter": 27,
        "hasPrmSection": false,
        "standingPlacesSecondClass": 133,
        "standingPlacesFirstClass": 0,
        "seatsCoupeSecondClass": 0,
        "seatsCoupeFirstClass": 0,
        "seatsSecondClass": 140,
        "seatsFirstClass": 0,
        "isDrivingCabinetAvailable": false,
        "hasServiceSection": false,
        "withTowBar": false,
        "withTractionCoupling": false,
        "hasBikeSection": false,
        "isEquipedForMux": true,
        "isSecondClass": true,
        "isFirstClass": false,
        "materialSubTypeSuffix": "BUH",
        "withAutomaticCoupling": false,
        "subTypeDescriptionFr": "HV M6 BUH",
        "subTypeDescriptionNl": "HV M6 BUH",
        "canPassToNextUnit": true,
        "hasCameras": false,
        "hasRealTimeInformationScreens": false,
        "hasWasteBins": true,
        "hasSemiAutomaticInteriorDoors": true,
        "hasLuggageSection": false
      },
    ...
    ]
  }
]
```

**Comment by Bertware** [comment]
@jncn @ChristopheVersieux @pietercolpaert Can you come up with any specific needs for such an endpoint? As far as I'm concerned, the data can be structured similarly to the existing iRail endpoints. I prefer a separate endpoint to prevent this (extra) code from affecting the crucial core endpoints.

I'd definitely like an open-source set of train icons based on train type (which should be included in the internal API), so we can show the exact train constellation to the users. Swedish example: 
![SJ app train constellation](https://user-images.githubusercontent.com/1094793/66641757-a5b6bf00-ec1b-11e9-9173-caa88f418b59.png). The locomotive/carriages match in type with the real train, so you can see exactly which train it's going to be. In this case we'd need a CC-0 or CC-By set with icons, and some designers who are willing to help a hand. This also prevents all apps/websites/... doing the same thing over and over.

**Comment by SIMPYbe** [comment]
Some info:
If pullPushType == "R", this is a Push-Pull train with a locomotive and coaches.
If pullPushType == "X", this is a train with motor-units (electric or diesel). There you could consider to filter only one unit, instead of per coach. You can do this by check the tractionPosition parameter. Note: in case of a motor-unit, the orientation of the unit is unknown.
I made a small testpage: https://project.haltelink.be/treinsamenstelling

**Comment by Bertware** [comment]
Another example, with older rolling stock:
```
{
        "hasReadingLights": false,
        "hasFoldingSeats": true,
        "hasAssistanceButtons": true,
        "hasToilets": true,
        "hasTables": true,
        "hasSecondClassOutlets": true,
        "hasFirstClassOutlets": false,
        "hasHeating": true,
        "hasAirco": false,
        "typeDescriptionNl": "AM 80 Vehicle c",
        "tractionType": "AM/MR",
        "pullPushType": "X",
        "diagramServiceNumber": 15,
        "materialNumber": 395,
        "materialDiagram": "MK",
        "parentUicCode": "948808039507",
        "uicCode": "948808039515",
        "isLocoInUse": true,
        "inDirection": false,
        "tractionPosition": 1,
        "positionInComposition": 1,
        "materialSubTypeName": "AM80_c",
        "typeDescriptionFr": "AM 80 Vehicle c",
        "parentMaterialSubTypeName": "AM80",
        "hasPriorityPlaces": false,
        "brutoWeight": 59,
        "lengthInMeter": 25,
        "hasPrmSection": false,
        "standingPlacesSecondClass": 15,
        "standingPlacesFirstClass": 9,
        "seatsCoupeSecondClass": 0,
        "seatsCoupeFirstClass": 0,
        "seatsSecondClass": 40,
        "seatsFirstClass": 32,
        "isDrivingCabinetAvailable": false,
        "hasServiceSection": false,
        "withTowBar": false,
        "withTractionCoupling": false,
        "hasBikeSection": false,
        "isEquipedForMux": false,
        "isSecondClass": true,
        "isFirstClass": true,
        "materialSubTypeSuffix": " ",
        "withAutomaticCoupling": false,
        "subTypeDescriptionFr": "AM 80 Vehicle c",
        "subTypeDescriptionNl": "AM 80 Vehicle c",
        "canPassToNextUnit": false,
        "hasCameras": false,
        "hasRealTimeInformationScreens": false,
        "hasWasteBins": true,
        "hasSemiAutomaticInteriorDoors": true,
        "hasLuggageSection": false
      },
```
I'm a bit unsure on how we can handle this best. On one side I'm sure someone will come up with a cool project that uses all this data, on the other hand it's a lot of data to handle (if we want to ensure the iRail output stays the same when the source data changes).


*Suggestion*: We include the most important data, and ensure it's always there. We add a query parameter to include the other data as raw data, as-is without guarantees, for those who want to use it. We offer one additional field which we calculate to give a clear and consistent indication on the traction type (as @SIMPYbe mentioned above) and the material type (to determine which visualisation should be shown)

The following fields seem the most important to me and should always be included, even if the source data would change its format:
```
        "hasToilets": true,
        "hasTables": true,
        "hasSecondClassOutlets": true,
        "hasFirstClassOutlets": false,
        "hasHeating": true,
        "hasAirco": false,
        "materialNumber": 395,
        "tractionType": "AM/MR",
        "canPassToNextUnit": false,
        "standingPlacesSecondClass": 15,
        "standingPlacesFirstClass": 9,
        "seatsCoupeSecondClass": 0,
        "seatsCoupeFirstClass": 0,
        "seatsSecondClass": 40,
        "seatsFirstClass": 32,
        "lengthInMeter": 25,
        "hasSemiAutomaticInteriorDoors": true,
        "hasLuggageSection": false
        "materialSubTypeName": "AM80_c"
```
Other data could then still be published, but without guarantees. The most critical work (type detection) could be done on the iRail server and added as a computed field, so end users don't need to handle all the different train types (who would be writing the same code over and over again). Would this work, or am I missing downsides to this approach?

**Comment by Bertware** [comment]
An endpoint for testing and evaluation is available at https://staging.api.irail.be/composition.php?id=ic2021&format=json

Data is cached for one hour for now, as this isn't something that changes so often.

**Comment by jncn** [comment]
Amazing, thanks @Bertware! This looks really promising, looking forward to use it in Railer. 
Unfortunately the staging link gives a 404 now.

**Comment by Bertware** [comment]
The staging link gives a 404 because that specific train number isn't in use at this moment. Search a train which is driving right now and it should work. If you get a 404 for a train you know is driving, try again (the nmbs' endpoint doesn't always answer)

**Comment by Bertware** [comment]
@jncn @ChristopheVersieux @DylanVanAssche feedback?

**Comment by DylanVanAssche** [comment]
I really like this composition API. It can be useful for filtering routes too on airco for example. In the summer, nobody likes a route without airco :)
Providing the most important fields is fine by me, I would also provide the bike section. You never know upfront where you need to wait on the platform with your bike. I did it in the past and it's like a slot machine :) 
- Beginning of the train
- In the middle of the train
- At the end of the train
- No bike section at all

**Comment by Bertware** [comment]
We can include the bike position in the "important fields" category :) The `&data=all` flag includes all raw data (without guarantees!) to allow anyone to experiment with all data, exactly for this case where we can't really know beforehand which fields will become important. I'm not sure on how we can include accessibility data in the best way, at this moment I feel like that's something the end user can recognize if the apps show drawings of the exact carriage types. There doesn't seem to be a flag indicating whether or not a carriage is wheelchair-friendly (probably because none of the trains really are at this moment)

https://staging.api.irail.be/composition.php?id=ic538&format=json&data=all

**Comment by Bertware** [comment]
We sometimes receive an invalid station UIC code: 8829025

```
    "ptCarFrom": {
      "id": 220,
      "uicCode": "8829025",
      "fromName": "Bruxell-Midi - Brussel-Zuid"
    },
    "ptCarTo": {
      "id": 1838,
      "uicCode": "8829025",
      "fromName": "Breda - Breda"
    },
    "confirmedBy": "Planning",
```

This may or may not be related to the "planning" status

**Comment by Bertware** [comment]
Vehicle types are still hard to resolve to the right classes as used by the [MlgTraffic ](http://www.mlgtraffic.net/SNCB_E_VM.xml#) image set. I'm wondering if we should create our own version of the image set where we rename the files to match the NMBS data (original data is CC BY-NC-SA), or if we should map the NMBS types in the iRail API to also output the right image name (for example BU -> B/Bd/ABd). Any thoughts or comments? @pietercolpaert ?

**Comment by Bertware** [comment]
Liege-Guillemins is reffered to by UIC code 008847159, which is wrong. In fact, most of their UIC codes are wrong. They might be reffering to a workplace?

**Comment by ChristopheVersieux** [comment]
Waouh, that's such an awesome feature (Sorry I am a little bit late, I had a very busy end of year)
I will gladly implement these features in the app!

1) I don't know if that's linked, but would be awesome to get a train type icon inside connections.php. would be more convenient for the user, instead of making separate requests for every connection returned. I don't see how apps can be broken with extra JSON fields (I hope)

2) I need more time to process all this data but the vehicle detail will definitely be implemented!!! 

Thanks for your hard work

**Comment by ChristopheVersieux** [comment]
First remarks after a quick play with the API, tested with: https://staging.api.irail.be/composition.php?id=L4662&format=json&data=all

1) In the JSON structure, I see composition twice: (**composition**/segments/segment[]/**composition**/units). Not really an issue, but will cause issues using some parsers. 

2) For the 2 segments, origin and destinations are the same (mons for segment 1 and Tournai for Segment 2) while https://trainmapjs.azureedge.net/data/composition/4662 has a Quevy/Mons and Tounai/Mons)

Maybe I missed something in previous comments regarding these remarks?

**Comment by Bertware** [comment]
Hi Christophe,

No pressure, just wanted to get some feedback before we "finish" the new endpoint, it's easier to change things before everyone has an implementation based on it.  

- The composition endpoint is completely separated from the other endpoints, and it even has a completely other datasource (NMBS trainmap, while the other data is based on website/app data). Adding it to the connections endpoint has the same drawbacks as making separate request, because the connections endpoint would have to make the same requests, making the connections endpoint slower. The client could fetch this data, which has the benefit that the client gets all composition data (instead of only an icon), meaning the developer can choose how to represent the data (text, icon, complete image, ...).

- I didn't think of that, do you have a suggestion on how we could rename one of those fields? The first is the endpoint name (composition fits perfectly there), the second is the composition for certain "legs", where it's also correct. It should be possible for clients to "hint" the right types in the class structure, so I'd personally prefer to continue with the current naming, unless it really is a blocker for common parsers.

- While the trainmap source shows different station names, it uses the same UIC code. This is a bug on the NMBS/SNCB side (which I don't think we can report, it's an internal API and doesn't seem bother them). We're using the UIC codes as names are quite hard to parse, but this also means we get incorrect information there.  This might be an issue we need to resolve as there might be trains that split/join/switch direction, meaning they'd have a different composition. 

I've tried to look around and verify the data in the last few days when taking a train, and it seems that while the composition is correct (right train in front, right vehicle numbers, ...) the orientation of AM/MS trains can be incorrect in the NMBS source data. This means that it can show first class/bike/... carriages in the wrong position for these trains. Definitely not always the case, but it can cause frustration with users if it's wrong just when they rely on it.

I currently implemented it in HyperRail to test/experiment with it, and based on those tests I'd say
- It's crucial to support "404" responses when no composition can be found. This is the case when the train hasn't left the first station, or when it has arrived at its destination. 
- The availability of toilets doesn't always mean that carriage has a toilet. It means you can reach a toilet from that carriage, so often all carriages are marked as "toilet: 1".
- Mapping carriages (M6 BHDX, M6 AAX, ...) to images can be hard, but it's definitely doable. AM/MR trains are marked with A/B/C/D instead of exact types. in this case, you need to combine knowledge (an AM08 has 3 carriages, meaning A and C are of the BDX type and B is a B type, while an AM75 contains 4 carriages, A and D are ABDX/BDX based on first class seats, B and C are B type, ...)
![2019-12-30 14 04 49](https://user-images.githubusercontent.com/1094793/71583124-6f4b5a80-2b0d-11ea-94f3-4d7bf9cd6d11.png)
![2019-12-29 19 01 41](https://user-images.githubusercontent.com/1094793/71583125-6f4b5a80-2b0d-11ea-8fea-24bcf31fa812.png)
